rfcs/generalized-docstrings.md

```diff
@@ -0,0 +1,192 @@
+after lexing, and documentation comments (of the form `(** ... *)`) are
+collected and attached to the parsetree as attributes identified with
+`ocaml.doc`.
+
+This RFC proposes to add new kinds of documentation comments, distinguished
+syntactically from the usual documentation comments available today.
+
+The concrete syntax would be as follows:
+
+```
+(** ... *)  : a usual doc string
+(*@ ... *)  : a doc string of kind "@"
+(*! ... *)  : a doc string of kind "!"
+...
+```
+
+i.e. allow a (reasonable) range of predefined characters as a "kind" marker
+after the opening comment tag `(*`.
+
+Then, different kinds of documentation comments would be elaborated to different
+attributes in the parsetree. Since `(** ... *)` produces attributes `ocaml.doc`,
+`(*@ ... *)` would produce attributes identified with e.g. `ocaml.doc_at`, etc.
+
+## Motivation
+
+This proposal is motivated by work as part of the [Vocal
+project](https://vocal.lri.fr/) ([github
+repo](https://github.com/vocal-project/vocal)).
+
+One goal of the project is to design a specification language for OCaml
+libraries, named Gospel, and taking inspiration in particular from
+[JML](https://en.wikipedia.org/wiki/Java_Modeling_Language).
+
+For that purpose, we want to attach to the items of a signature not only
+human-readable docstrings (between `(** ... *)`, to be processed by ocamldoc),
+but also machine-readable formal specifications, to be processed by a dedicated
+tool including a dedicated parser. It is convenient to be able to write these
+specifications in a dedicated comment block, between `(*@ ... *)`.
+
+Note that we *do not want* our specification-comment-blocks to be picked up by
+ocamldoc/odoc as standard docstrings (initially they would be simply ignored;
+and then we could imagine implementing dedicated support for them in odoc).
+
+One example from the vocal repo, specifying (both formally and informally) the
+`create` function of a
+[Vector](https://github.com/vocal-project/vocal/blob/master/src/Vector.mli)
+module.
+
+```ocaml
+val create: ?capacity:int -> dummy:'a -> 'a t
+(** [create] returns a fresh vector of length [0].
+   All the elements of this new vector are initially
+   physically equal to [dummy] (in the sense of the [==] predicate).
+   When [capacity] is omitted, it defaults to 0. *)
+(*@ a = create ?capacity ~dummy
+      requires let capacity = match capacity with
+                 | None -> 0 | Some c -> c in
+               0 <= capacity <= Sys.max_array_length
+      ensures  length a.view = 0 *)
+```
+
+We believe that it makes sense to distinguish syntactically between docstrings
+"for humans", and formal specifications: they correspond to two different levels
+of abstraction, and will be processed by different tools.
+
+Instead of introducing new syntax, we could require the use to directly use the
+syntax of attributes, and instead write something like (note that the user now
+needs to worry about proper escaping between the `"`):
```

Comment:
We could use `{| |}` to delimit strings and don't worry about escaping.

Reply:
also, ppx payload is not required to be a string and you can employ OCaml syntax for annotations. Not sure whether it is a good idea or not, just a food for thought.